🎛️ Migration of `tabs.js` to Stimulus `TabsController` (`w-tabs`)

lb- opened the pull request:
## Description

- Fixes #11501
- Migrates the current `tabs.js` to a new Stimulus controller `TabsController` with the identifier `w-tabs`
- This enhancement includes full unit tests for the entire behaviour, fixes for edge case keyboard control within tabs and migrates the code over to TypeScript.
- We no longer need to manually `initTabs` wherever the tabs may appear or change (e.g. chooser modals), instead just adding or mutating the DOM will be enough to ensure the tabs work wherever they are used.

> This PR is a rebase and rework of the awesome progress by @the-r3aper7 on #11542

## Details

- This PR contains granular commits to more easily review, it contains an intentional commit to 'rename' the original tabs file to the new Stimulus file so that the file history is preserved.
- Fixes a bug (found during testing) where using `cmd+left-arrow` or `cmd+right-arrow` on a tab will conflict with the keyboard handler and browser back, causing no change.
- Instead of class changes & transitions being tied together and no ability to change the 'active class', this now isolates the transition timing and the class changes for better future evolution of tabs.
- Instead of history/URL syncing being opt out, this is now opt in, to make it easier to use nested tabs.
- Adds an `after` util, basically a rebuild of [Lodash `after`](https://lodash.com/docs/4.17.15#after) to make it easier to ensure that a function will run only after `n` calls.
- Updates the events dispatched to use the built in consistent naming for Stimulus. Instead of `switch` (for tab selection) and `wagtail:tab-changed` (for changes, dispatched on document), these are now `w-tabs:selected` (dispatching on all tab/triggers) and `w-tabs:changed` (dispatches on the actual tabs container). Making these events more usable & predictable.
- Note that no animation is in use today, however, as it was in the code and its been ported to the new controller.
- Enhances the Styleguide tabs examples, with some slightly refined styling and two discrete examples to make it easier to understand how tabs can be used.
- The other enhancement is we get a set of properly formed errors (without breaking other components) if the DOM structure, including accessibility items, is not correct. Before, these would kind of fail silently or not be flagged at all.
- Full unit tests (only coverage gaps are actually unused code, only there to keep TypeScript happy), test coverage comparison below.

| File                        | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s                                          |
| --------------------------- | ------- | -------- | ------- | ------- | ---------------------------------------------------------- |
| TabsController.ts (this PR) | 97.89   | 93.75    | 94.59   | 100     | 169-170,254,294                                            |
| tabs.js (current/`main`)    | 57.34   | 52.94    | 50      | 56.73   | ...101,105,120,147,151,177-196,215-219,227-268,307-337,367 |

## Testing

### Tabs are used in the following locations

| Usage                                            | Screenshot                                                                                                                                         |
| ------------------------------------------------ | -------------------------------------------------------------------------------------------------------------------------------------------------- |
| `TabbedInterface` (page/snippet edit & creation) | ![Screenshot 2024-11-14 at 8 23 41 PM](https://github.com/user-attachments/assets/93763cd0-fe96-4e3a-9752-6622864e7192)                            |
| Account editing                                  | ![Screenshot 2024-11-14 at 8 24 50 PM](https://github.com/user-attachments/assets/ee79695b-ba83-47ef-8965-2a25d8de0654)                            |
| Choosers (with tabs, e.g. image)                 | ![Screenshot 2024-11-14 at 8 25 50 PM](https://github.com/user-attachments/assets/68d07f59-41b8-4afb-b882-c633b1fe4951)                            |
| Workflow history                                 | ![Screenshot 2024-11-14 at 8 26 35 PM](https://github.com/user-attachments/assets/52f8c901-14e0-4212-9b97-4776e60bba8d)                            |
| Workflow task chooser                            | ![Screenshot 2024-11-14 at 8 27 13 PM](https://github.com/user-attachments/assets/76472693-8f04-4f00-8237-d6ea16634a2c)                            |
| User edit/create views                           | ![Screenshot 2024-11-14 at 8 28 00 PM](https://github.com/user-attachments/assets/e7dfa9d3-5c9e-48e1-a6ce-698f9eed3565)                            |
| New Stimulus Storybook                           | <img width="1048" alt="Screenshot 2024-11-14 at 9 07 39 PM" src="https://github.com/user-attachments/assets/a59cd8ea-1003-4345-b803-f0bd3f532077"> |

### Items tested

- Keyboard control (when focused on a tab), left/right/home/end (including 'looping' around tabs).
- URL updates, including back/forward browser changing tab if there's been a change after load.
- Deep linking into content within tabs.
- Comments sidebar showing items based on the current tab, updating comments when tabs change.
- Minimap listing reflecting the currently visible tab items.
- Storybook.
- Animation styling (pretty basic though).
- Manual testing of malformed HTML to ensure console errors show.

### Browsers tested

- Firefox 132 on MacOS 14.5
- Safari 17.5 on MacOS 14.5
- Chrome 130 on MacOS 14.5

## Release notes

> Changelog, please ensure we include @the-r3aper7 who worked on this extensively and put together a substantial amount of the unit tests and change discovery work.

We will likely want to call out this change, even though it's undocumented and it would have been quite difficult to use the existing tabs in any scenario beyond initial HTML on load.

## Upgrade considerations - changes to undocumented internals

The previous Tabs frontend implementation has changed drastically, this has now been migrated to Stimulus.

Any existing usage of `data-tabs` will no longer work, here's a simple summary of the new approach.

These are subject to change and are still not officially documented, please add a thumbs up or comment to [#8408](https://github.com/wagtail/wagtail/issues/8408) if a formalised Python or JavaScript API for this would be used.

| Item      | Usage                                     | Old                                                                 | New                                                                                                                                                                            |
| --------- | ----------------------------------------- | ------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| Attribute | Root tab container div                    | `data-tabs`                                                         | `data-controller="w-tabs"`                                                                                                                                                     |
| Attribute | Tab trigger (tab)                         | `role="tab"` only                                                   | `role="tab"` & `data-w-tabs-target="trigger" data-action="w-tabs#select:prevent"`                                                                                              |
| Attribute | Non-tab ad-hoc triggers (as a link)       | `href="#tab-label-id" data-trigger`                                 | `href="#the-tab-id" data-w-tabs-target="trigger" data-action="w-tabs#select:prevent" data-w-tabs-focus-param="true"` (focus is optional, but recommended)                      |
| Attribute | Non-tab ad-hoc triggers (as a button)     | N/A - was not supported                                             | `type="button" data-w-tabs-target="trigger" data-action="w-tabs#select" data-w-tabs-id-param="the-tab-id" data-w-tabs-focus-param="true"` (focus is optional, but recommended) |
| Attribute | Opt in/out of URL sync, on tabs container | `data-tabs-disable-url` (opt out)                                   | `data-action="popstate@window->w-tabs#select" data-w-tabs-use-location-value="true"` (opt in)                                                                                  |
| Attribute | Enable animation                          | `data-tabs-animate` (hard-coded 150ms)                              | `data-w-tabs-transition-value="150"`                                                                                                                                           |
| Attribute | Add a class when a panel becomes active   | `data-tabs-animate` (hard-coded 'animate-in' class)                 | `data-w-tabs-active-class="animate-in"`                                                                                                                                        |
| Attribute | Tab panel (content)                       | `role="tabpanel"` only                                              | `role="tabpanel"` & `data-w-tabs-target="panel"`                                                                                                                               |
| Attribute | Tablist container keyboard control        | `role="tablist"`                                                    | `role="tablist"` & `data-action="keydown.right->w-tabs#selectNext keydown.left->w-tabs#selectPrevious keydown.home->w-tabs#selectFirst keydown.end->w-tabs#selectLast"`        |
| Event     | A tab is selected                         | `'switch'` (dispatched on the `role=tablist`, does not bubble)      | `'w-tabs:selected'` (dispatched on the triggers for the tab, does bubble)                                                                                                      |
| Event     | A change to the active panel              | `'wagtail:tab-changed'` (dispatched on `document`, does not bubble) | `'w-tabs:changed'` (dispatched on the tabs container, does bubble)                                                                                                             |

For a full HTML example, install the [](styleguide) and see the tabs examples there.

> Reminder, the link just above will not work in GitHub, only when used in the docs.

## Future enhancements

This PR unlocks a few items for enhancement and further development.

1. #8408 - This would be a great candidate for work after this PR.
2. #8483 - We would be in a great position to do this now, with a solid set of unit tests to build around.
3. https://github.com/wagtail/wagtail/issues/12128 - Easier enhancement for setting the tab on our forced focus utility, so we can focus on the first error with a button click.
4. Equalized heights of tabs, it would be great to not have content jump around when switching tabs, we could enhance this controller to ensure that the tab content heights are always set to the 'tallest' tab element. See code snippet below for another codebase that's done this.
   https://github.com/Sub-Xaero/stimulus-library/blob/24b3b8e6635d24cc8bb66865478e4df8c483f24b/packages/controllers/src/visual/tabs_controller.ts#L88-L110

lb- commented:
Rebased and aligned JSDoc with other conventions

lb- commented:
This PR may be an enabler for https://github.com/wagtail/wagtail/issues/12128 - see comment https://github.com/wagtail/wagtail/pull/12739#pullrequestreview-2728027497 

@laymonage  - pinging for a review if you get the chance.

zerolab commented on wagtail/admin/templates/wagtailadmin/account/account.html:
```diff
@@ -2,9 +2,9 @@
+    <div class="w-tabs" data-controller="w-tabs" data-action="popstate@window->w-tabs#select" data-w-tabs-use-location-value="true">
         <div class="w-tabs__wrapper">
-            <div role="tablist" class="w-tabs__list">
+            <div role="tablist" class="w-tabs__list" data-action="keydown.right->w-tabs#selectNext keydown.left->w-tabs#selectPrevious keydown.home->w-tabs#selectFirst keydown.end->w-tabs#selectLast">
```
question: is there a way to make `data-action="keydown.right->w-tabs#selectNext keydown.left->w-tabs#selectPrevious keydown.home->w-tabs#selectFirst keydown.end->w-tabs#selectLast"` a default behaviour, then allow for overrides?

Or, perhaps we need better documentation. I am thinking with my package maintainer hat on

lb- replied:
I agree this is a bit verbose at first glance, it's intentional though and it's how keyboard interactions get written for other Stimulus usage. 

In regards to extending though, that could be done a few ways in the future, mainly though we should be bringing in these disparate tabs html template partials into one place. 

I didn't want this PR to get too large and am hoping, once merged, we'll be able to make proper progress on https://github.com/wagtail/wagtail/issues/8408

The idea being that developers, and package maintainers, can leverage shared templates and a common Python API for Tabs. Similar to Panels.

The other benefit of having this in HTML is that we can more easily support changed behaviour based on locale or anything else that's a template concern.

Happy to abstract this out to the JS though, it just means we have to do special handling for any other `data-action` that may also be set on the HTML.

lb- replied:
Epic. Thanks for trying this out.

laymonage commented on client/src/controllers/TabsController.ts:
```diff
@@ -0,0 +1,553 @@
+    this.triggerTargetConnected = resetTabs;
+    this.triggerTargetDisconnected = resetTabs;
+
+    // Ensure the syncLocation runs only the second time so we do not update the URL/history on load
+    this.syncLocation = after(this.syncLocation.bind(this), 1);
```
I wonder why this works despite being done at the end of `connect`? I thought all the previous initialisation code would've already called `syncLocation()` if we want to prevent that from happening on first load.

Also, I'm not sure what to feel about this `after` function. Didn't know such a pattern exist. It's interesting and I can see how it can be useful. Although if our use case is for "preventing it from being called until a later point (i.e. we've done the initial setup/tab selection)", I think I'm more inclined to having a boolean e.g. `this.ready` that's set to `true` at the point in code where we're happy for the method to run, rather than relying on the number of times it's called. Then, `if (!this.ready)`, do an early return in the method (or avoid calling it).

This is my personal take though, I'm not against`after` if it helps make things easier.

lb- commented:
@laymonage thank you! This is an incredibly helpful feedback and review, I'll let you know when it's ready for another.

lb- commented on client/src/controllers/TabsController.ts:
```diff
@@ -0,0 +1,553 @@
+  afterTransition(element: HTMLElement) {
+    const transitionValue = this.transitionValue || 0;
+
+    if (transitionValue <= 0) return immediatePromiseLike(element);
```
Yes, this is a good point, originally I thought this may improve the response time for non-animation scenarios but now that I have reviewed I can see this will work the same without this complexity.

Thank you.

lb- commented on client/src/controllers/TabsController.ts:
```diff
@@ -0,0 +1,553 @@
+  };
+
+  /** The `id` of the panel (not the tab/trigger) that's currently active. */
+  declare activePanelIdValue: string | null;
```
Good find, fixed.

lb- commented on client/src/controllers/TabsController.ts:
```diff
@@ -0,0 +1,553 @@
+        this.activePanelIdValue = panelId;
+        this.tabs
+          .filter((tab) => tab.getAttribute('aria-controls') === panelId)
+          .forEach((tab) => tab.focus());
```
Yep, good call on this. I have taken a different tact but achieved the same goal of focusing only on the right tab and only those that are valid tabs.

lb- commented on client/src/controllers/TabsController.ts:
```diff
@@ -0,0 +1,553 @@
+        this.activePanelIdValue = panelId;
+        return panelId;
+      }
+    }
```
Yep, I took a different approach but cleaned this up a bit and also ensured this method is a bit more flexible for the index-like select and the external trigger, with better focus management now centralised into this method.

lb- commented on client/src/controllers/TabsController.ts:
```diff
@@ -0,0 +1,553 @@
+   * Set the first tab's panel to the active panel.
+   */
+  selectFirst(event?: Event) {
+    const tab = this.tabs[0];
```
Defensive tab selection checks are now in all select methods.

lb- commented:
@laymonage  OK, this is now good to go for another review, hopefully I have addressed all the concerns.

I have added a wider set of unit tests for the tabindex management and the focus management, especially for external triggers.

lb- commented on client/src/controllers/TabsController.ts:
```diff
@@ -243,13 +241,18 @@ export class TabsController extends Controller {
   async afterTransition(element: HTMLElement) {
-    const transitionValue = this.transitionValue || 0;
+    const { transitionDelay, transitionDuration } = getComputedStyle(element);
+    // Browsers normalize the value in seconds, convert to milliseconds
+    // for use in setTimeout.
```
Yep. It should return seconds as it's the canonical unit for serialisation. 

Bit of a rabbit hole but here's the links. 

https://developer.mozilla.org/en-US/docs/Web/CSS/time
https://drafts.csswg.org/css-values/#time
https://drafts.csswg.org/css-values/#canonical-unit
https://www.w3.org/TR/css-values-4/#time-value
https://www.w3.org/TR/cssom-1/#serializing-css-values

lb- replied:
Performance should be ok, as long as we aren't thrashing the Dom with these kinds of queries 100s of times a second.

laymonage replied:
Thanks for the pointers, good to know!

lb- commented on wagtail/contrib/styleguide/templates/wagtailstyleguide/base.html:
```diff
@@ -444,20 +444,20 @@ <h3>Tabs</h3>
             </div>
             <div class="tab-content">
-                <section class="w-min-h-24" id="tab-basic-tab-1" role="tabpanel" aria-labelledby="tab-label-basic-tab-1" data-w-tabs-target="panel">
+                <section class="w-min-h-24 w-tabs__panel" id="tab-basic-tab-1" role="tabpanel" aria-labelledby="tab-label-basic-tab-1" data-w-tabs-target="panel">
```
Thanks. I must have missed that one.

lb- commented on client/src/controllers/TabsController.ts:
```diff
@@ -72,8 +71,6 @@ export class TabsController extends Controller {
   /** Tab content panels, with role='tabpanel', showing the content for each tab. */
   declare readonly panelTargets: HTMLElement[];
-  /** The time, in milliseconds, for the tab content to transition in and out. If zero or negative, no transition will take place. */
-  declare readonly transitionValue: number;
```
Thanks @laymonage I agree with your direction here. I think I was just trying to make the controller more flexible and allow for a max duration.

However, if that need arises we can add it later.

Thanks again for your input on this.

lb- commented on client/src/components/Minimap/index.tsx:
```diff
@@ -60,7 +60,7 @@ const createMinimapLink = (
 const renderMinimap = (container: HTMLElement) => {
   let anchorsContainer: HTMLElement = document.body;
-  const tabs = document.querySelector('[data-tabs]');
+  const tabs = document.querySelector('[data-controller~="w-tabs"]');
```
I like the approach taken in the commit.

I think we may need to review the minimap code at some point and see what other approaches we can take, this appears to work well enough for now.

lb- commented:
Thanks @laymonage  - I am happy with your changes as applied to the branch. Thanks for the thorough review of this code and testing. Thanks also for your various rounds of testing on this @zerolab 

It would be nice if we could clean up these commits into the original commit set, the 'cleanup' commit & the CSS transition commit could be merged into the main tabs controller migration commit. The others make sense as isolated commits though.

I think we should just remove `this.ready` completely if it's not used. I do, however, think we should consider a generic approach to `this.ready` in an abstract class though, it's a pretty helpful thing. That can come later though.

Thanks for fixing up the `tabbed_interface` stories, sorry I missed that one.

Let me know if you want me to do the final rebase & merge.

laymonage reviewed (approved):
Awesome work, thanks @lb- and @saisrikardumpeti! I've `fixup`'d the commits and removed `this.ready`. Will merge when CI passes.

lb- commented:
Thanks again @laymonage
